### PR TITLE
FailureDetector (safe version)

### PR DESCRIPTION
**Pull request by bresch**
Save version of #10179

This PR has no "dangerous" changes as it only sets a `attitude_failure` flag in `vehicle_status` but doesn't activate a failsafe action.

This PR needs to go in master for testing and is completely invisible to the user.

The failsafe triggering part will come in a future PR.

FYI @LorenzMeier @dagar @JohannesBrand

### CODE REVIEW

File: src/modules/commander/commander_params.c [hunk 1]
```diff
@@ -749,6 +749,30 @@ PARAM_DEFINE_FLOAT(COM_POS_FS_EPV, 10);
...
+ * FailureDetector Max Roll
+ *
+ * Maximum roll angle before FailureDetector triggers the attitude_failure flag
+ * Does not affect the behavior of the vehicle for now; only for logging
+ * @min 0
+ * @max 180
+ * @unit degrees
+ * @group Commander
+ */
+PARAM_DEFINE_INT32(COM_FAIL_R, 60);
```
Comment (dagar) [inline src/modules/commander/commander_params.c]: If you foresee the list of FailureDetector parameters growing I'd break it off into a new group (and param file) now.

File: src/modules/commander/Commander.cpp [hunk 2]
```diff
@@ -2385,6 +2386,32 @@ Commander::run()
 			}
 		}
 
+
+		/* Check for failure detector status */
+		if (armed.armed) {
```
Comment (dagar) [inline src/modules/commander/Commander.cpp]: && vehicle_control_mode.flag_control_attitude_enabled?

**Comment by dagar** [comment]
Shouldn't this be based on the error rather than absolute angle?

**Comment by bresch** [comment]
> Shouldn't this be based on the error rather than absolute angle?

Yes, but he problem are the setpoint jumps...

**Comment by dagar** [comment]
> > Shouldn't this be based on the error rather than absolute angle?
> 
> Yes, but he problem are the setpoint jumps...

I know, but I'm talking about more of a usability concern. The user shouldn't be able to configure this such that it triggers under normal operation. This is one of the reasons I had originally wondered about the controller being responsible for raising the error condition, and commander the actual triggering.

**Comment by bresch** [comment]
@dagar Absolutely, we might add a check to make sure the parameters are larger than the max roll/pitch values of the controller. Basically, with this PR, I just want to introduce the general skeleton of the failure detector. The logic inside is really simple and will be improved later.
It will also be a good reason to fix the setpoint jumps :)

File: src/modules/commander/Commander.cpp (hunk 2 above)
Comment (bresch) [reply src/modules/commander/Commander.cpp]: I would like to leave it like this as it is just an absolute attitude check and not an attitude error check (for now).

**Comment by dagar** [comment]
Let me start by saying the PR in its current form looks good and I'm happy to merge it once we get it to fit.

> we might add a check to make sure the parameters are larger than the max roll/pitch values of the controller. Basically, with this PR, I just want to introduce the general skeleton of the failure detector. 

Yes exactly, but my (future) concern is that this piece within commander then becomes dependent on attitude controller internals (particular parameters for each vehicle and how they're used in different contexts). Which is one of the reasons I get hung up on the architecture now for something so simple.

File: msg/vehicle_status.msg [hunk 3]
```diff
@@ -67,7 +74,7 @@ bool high_latency_data_link_active 		# all low latency datalinks to GCS lost
 uint8 data_link_lost_counter			# counts unique data link lost events
 bool engine_failure				# Set to true if an engine failure is detected
 bool mission_failure				# Set to true if mission could not continue/finish
-bool attitude_failure 				# Set to true if attitude failure has been detected
+uint8 failure_detector_status			# Bitmask containing FailureDetector status [0, 0, 0, 0, 0, FAILURE_ALT, FAILURE_PITCH, FAILURE_ROLL]
```
Comment (dagar) [inline msg/vehicle_status.msg]: This is fine, but you can feel free to use bools here if you want. The message is low rate so the space savings are negligible. For code though it requires more work to get or set a bit, and then for log analysis it ends up being a mess because the constants are lost. 

For example estimator_status bitfields (https://github.com/PX4/Firmware/blob/master/msg/estimator_status.msg#L25) ended up being a poor choice. These constants have been manually duplicated across both ecl and Firmware, the ecl analysis python scripts under Tools, in Flight review, etc. They're also fairly inaccessible with flightplot.

tl;dr - booleans + publish on change is your friend

**Comment by LorenzMeier** [comment]
Could you please rebase? Thanks!

**Comment by bresch** [comment]
Rebased on master

**Comment by bresch** [comment]
@dagar Is it fine to merge now?

File: src/modules/commander/failure_detector/failure_detector_params.c [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+/**
+ * FailureDetector Max Roll
+ *
+ * Maximum roll angle before FailureDetector triggers the attitude_failure flag
+ * Does not affect the behavior of the vehicle for now; only for logging
+ * @min 0
+ * @max 180
+ * @unit degrees
+ * @group FailureDetector
```
Comment (dagar) [inline src/modules/commander/failure_detector/failure_detector_params.c]: This is the group shown in the QGC parameter tree and contain spaces.

eg  `@group FW L1 Control`

**Comment by dagar** [comment]
That's odd, github won't let me "Rebase and merge", but I could still "Squash and merge".

If you're fine with this going into master as single merged commit we can merge now, otherwise please rebase on master and we'll try again.

**Comment by bresch** [comment]
@dagar  Rebased on master! :) Hope it works now

**Comment by dagar** [comment]
Thanks, everything looks good now.
